xtask/src/task_phd.rs

```diff
@@ -0,0 +1,349 @@
+            args::CRUCIBLE_COMMIT => crucible_commit,
+            args::ARTIFACTS_TOML => artifacts_toml,
+        }
+    }
+
+    let propolis_local_path = match propolis_local_path {
+        Some(path) => path.into(),
+        None => {
+            let bin = build_bin("propolis-server")?;
+            let path = bin
+                .path()
+                .try_into()
+                .context("Propolis server path is not UTF-8")?;
+            relativize(path).to_path_buf()
+        }
+    };
+
+    let meta = cargo_metadata::MetadataCommand::new()
+        .no_deps()
+        .exec()
+        .context("Failed to run cargo metadata")?;
+    let phd_dir = relativize(&meta.target_directory).join("phd");
+    let artifact_dir = phd_dir.join("artifacts");
+    mkdir(&artifact_dir, "artifact directory")?;
+
+    let tmp_dir = {
+        let mut tmp_dir = phd_dir.join("tmp");
+        let now = time::SystemTime::now();
+        delete_old_tmps(&tmp_dir, now)?;
+        tmp_dir.push(
+            now.duration_since(time::UNIX_EPOCH).unwrap().as_secs().to_string(),
+        );
+        tmp_dir
+    };
+
+    mkdir(&tmp_dir, "temp directory")?;
+
+    let artifacts_toml =
+        artifacts_toml.map(Utf8PathBuf::from).unwrap_or_else(|| {
+            // if there's no explicitly overridden `artifacts.toml` path,
+            // determine the default one from the workspace path.
+            relativize(&meta.workspace_root)
+                .join("phd-tests")
+                .join("artifacts.toml")
+        });
+
+    if artifacts_toml.exists() {
+        cargo_log!("Found", "artifacts.toml at `{artifacts_toml}`")
+    } else {
+        anyhow::bail!("Missing artifacts config `{artifacts_toml}`!");
+    }
+
+    let status = run_exit_code(
+        phd_runner
+            .command()
+            .arg(args::RUN)
+            .arg(args::PROPOLIS_CMD)
+            .arg(&propolis_local_path)
+            .arg(args::CRUCIBLE_COMMIT)
+            .arg(crucible_commit.unwrap_or("auto"))
+            .arg(args::PROPOLIS_BASE)
+            .arg(propolis_base_branch.unwrap_or("master"))
+            .arg(args::ARTIFACTS_TOML)
+            .arg(&artifacts_toml)
+            .arg("--artifact-directory")
+            .arg(&artifact_dir)
+            .arg("--tmp-directory")
+            .arg(&tmp_dir)
+            .args(bonus_args),
+    )?;
+
+    std::process::exit(status);
+}
+
+mod args {
+    pub(super) const RUN: &str = "run";
+    pub(super) const PROPOLIS_CMD: &str = "--propolis-server-cmd";
+    pub(super) const PROPOLIS_BASE: &str = "--base-propolis-branch";
+    pub(super) const CRUCIBLE_COMMIT: &str = "--crucible-downstairs-commit";
+    pub(super) const ARTIFACTS_TOML: &str = "--artifact-toml-path";
+}
+
+fn build_bin(name: impl AsRef<str>) -> anyhow::Result<escargot::CargoRun> {
+    const PROFILE: &str = if cfg!(debug_assertions) {
+        "dev [unoptimized + debuginfo]"
+    } else {
+        "release [optimized]"
```

Comment:
How do you specify this with an xtask? (Do you have to manually invoke the xtask via `cargo run`?)

It's generally undesirable to build `phd-runner` with the `release` profile in this workspace, because that enables abort-on-panic, and the runner expects panics to unwind so it can catch them to report test failures. It might be nice to have an option to override the `current_release` setting here if we know we're building the runner itself.

Reply:
:+1:  will changee that!

Reply:
> How do you specify this with an xtask? (Do you have to manually invoke the xtask via `cargo run`?)

Yes, but we have a  cargo alias for xtasks, so you can type `cargo xtask ...` and get `cargo run` with some bonus flags:
https://github.com/oxidecomputer/propolis/blob/c596d09aa152ab8c8cf5086756ae86b16708d9b4/.cargo/config.toml#L1-L2

Currently, if you run the xtask using `cargo xtask`, you'll never be able to specify a release build, because the `cargo xtask` alias passes all arguments to the xtask binary, so if you type `cargo xtask phd --release`, the `--release` argument is passed _to the xtask binary_ rather than to `cargo`.

Reply:
as of 06e0fa60084af15b005ca9efff16df4cbf23c10a and 3e77e05df08e75c09680d3f2d14219677784a75d, we now always build `phd-runner` in debug mode, and i've added a `--release`/`-r` flag to build `propolis-server` in release mode (if you want to do that):

```console
$ cargo xtask phd run --release
   Compiling phd-runner
    Finished phd-runner dev [unoptimized + debuginfo] in 0.25s
   Compiling propolis-server
    Finished propolis-server release [optimized] in 87.05s
       Found existing artifact directory `target/phd/artifacts`
     Created temp directory `target/phd/tmp/1706310120`
       Found artifacts.toml at `phd-tests/artifacts.toml`
     Running target/x86_64-unknown-linux-gnu/debug/phd-runner run \
        --propolis-server-cmd target/x86_64-unknown-linux-gnu/release/propolis-server \
        --crucible-downstairs-commit auto \
        --base-propolis-branch master \
        --artifact-toml-path phd-tests/artifacts.toml \
        --artifact-directory target/phd/artifacts \
        --tmp-directory target/phd/tmp/1706310120
2024-01-26T23:02:00.279763Z  WARN phd_runner: 27: Failed to enable one or more kernel options, some tests may not work error=set_vmm_globals requires illumos
```

and, if you combine `--release` with a custom `propolis-server` path, you get a warning! :)

```console
:; cargo xtask phd run --release --propolis-server-cmd target/x86_64-unknown-linux-gnu/debug/propolis-server
   Compiling phd-runner
    Finished phd-runner dev [unoptimized + debuginfo] in 0.24s
  Overridden --propolis-server-cmd "target/x86_64-unknown-linux-gnu/debug/propolis-server"
warning: setting `--release` to build propolis-server in release mode does nothing if an existing propolis binary path is provided using `--propolis-server-cmd`
       Found existing artifact directory `target/phd/artifacts`
     Created temp directory `target/phd/tmp/1706310263`
       Found artifacts.toml at `phd-tests/artifacts.toml`
     Running target/x86_64-unknown-linux-gnu/debug/phd-runner run \
        --propolis-server-cmd target/x86_64-unknown-linux-gnu/debug/propolis-server \
        --crucible-downstairs-commit auto \
        --base-propolis-branch master \
        --artifact-toml-path phd-tests/artifacts.toml \
        --artifact-directory target/phd/artifacts \
        --tmp-directory target/phd/tmp/1706310263
2024-01-26T23:04:23.328977Z  WARN phd_runner: 27: Failed to enable one or more kernel options, some tests may not work error=set_vmm_globals requires illumos
```